### PR TITLE
Code Review

### PR DESCRIPTION
**Pull request by ps789**
# Overview

## Checklist
- [x] Interrogate has > 70% coverage
- [x] All tests have passed (new tests are made)
- [ ] Total test coverage > 75%
- [ ] New code test coverage > 80%
- [ ] Pass all linting checks

### CODE REVIEW

File: src/coalescenceml/artifact_store/base_artifact_store.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+            raise NotFoundError() from e
+
+    return inner_function
+
+
+class BaseArtifactStore(StackComponent, ABC):
+    """Base class for all CoalescenceML artifact stores.
+    Attributes:
+        path: The root path of the artifact store.
+    """
```
Comment (zhangtravis) [inline src/coalescenceml/artifact_store/base_artifact_store.py:46]: Good use of comments!

File: src/coalescenceml/artifacts/base_artifact.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    PRODUCER_PROPERTY_KEY,
+)
+
+
+DATATYPE_PROPERTY = Property(type=PropertyType.STRING)
+PRODUCER_PROPERTY = Property(type=PropertyType.STRING)
+
+
+class BaseArtifact(Artifact):
+    """ """
```
Comment (zhangtravis) [inline src/coalescenceml/artifacts/base_artifact.py:17]: Would be useful to explain what the difference between `BaseArtifact` and `BaseArtifactStore` is

File: src/coalescenceml/artifacts/metrics_artifact.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
+from coalescenceml.artifacts.base_artifact import BaseArtifact
+
+
+class MetricsArtifact(BaseArtifact):
+    """"""
```
Comment (zhangtravis) [inline src/coalescenceml/artifacts/metrics_artifact.py:5]: Artifact to keep track of the metrics?

File: src/coalescenceml/cli/integration.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+from coalescenceml.constants import console
+from coalescenceml.integrations.registry import integration_registry
+from coalescenceml.logger import get_logger
+
+
+logger = get_logger(__name__)
+
+
+@cli.group(help="Interact with the requirements of external integrations.")
+def integration() -> None:
```
Comment (zhangtravis) [inline src/coalescenceml/cli/integration.py:28]: What's the purpose of this method?

File: src/coalescenceml/cli/stack.py [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+from coalescenceml.constants import console
+from coalescenceml.directory import Directory
+from coalescenceml.enums import StackComponentFlavor
+from coalescenceml.stack import Stack
+from coalescenceml.stack.exceptions import ProvisioningError
+
+
+# Stacks
+@cli.group()
+def stack() -> None:
```
Comment (zhangtravis) [inline src/coalescenceml/cli/stack.py:17]: What is the purpose of this method?

File: src/coalescenceml/cli/integration.py (hunk 4 above)
Comment (zhangtravis) [inline src/coalescenceml/cli/integration.py:28]: What's the purpose of this method?

File: src/coalescenceml/artifacts/metrics_artifact.py (hunk 3 above)
Comment (zhangtravis) [inline src/coalescenceml/artifacts/metrics_artifact.py:5]: Artifact to keep track of metrics?

File: src/coalescenceml/artifacts/base_artifact.py (hunk 2 above)
Comment (zhangtravis) [inline src/coalescenceml/artifacts/base_artifact.py:17]: Would be useful to explain what the difference between `BaseArtifact` and `BaseArtifactStore` is

File: src/coalescenceml/artifact_store/base_artifact_store.py (hunk 1 above)
Comment (zhangtravis) [inline src/coalescenceml/artifact_store/base_artifact_store.py:46]: Good use of comments!

File: src/coalescenceml/integrations/tensorflow/step/tensorflow_trainstep.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+                        input_dim=len(x_train[0, :]),
+                        activation="relu",
+                    )
+                )
+            else:
+                model.add(Dense(layers[x], activation="relu"))
+        model.add(Dense(1, activation="sigmoid"))
+
+        model.compile(
+            loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"]
```
Comment (zhangtravis) [inline src/coalescenceml/integrations/tensorflow/step/tensorflow_trainstep.py:40]: Is the model only trained on binary crossentropy? What about for other losses and optimizers?

File: src/coalescenceml/orchestrator/__init__.py [hunk 7]
```diff
@@ -0,0 +1,11 @@
+"""
+## Orchestrator
+
+An orchestrator is a special kind of backend that manages the running of each
+step of the pipeline. Orchestrators administer the actual pipeline runs.
+"""
```
Comment (zhangtravis) [inline src/coalescenceml/orchestrator/__init__.py:6]: I think it would be useful if every major component has something similar to this so that those who don't understand the overall structure of COML (including me :')) gets a better understanding of how it works

File: src/coalescenceml/artifact_store/base_artifact_store.py [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+
+from coalescenceml.artifact_store.exceptions import ArtifactStoreInterfaceError
+from coalescenceml.enums import StackComponentFlavor
+from coalescenceml.stack import StackComponent
+
+
+PathType = Union[bytes, str]
+
+
+# Bullshit hack b/c of TFX....
```
Comment (ew423) [inline src/coalescenceml/artifact_store/base_artifact_store.py:27]: it do be like that

File: src/coalescenceml/artifact_store/base_artifact_store.py [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+        super(BaseArtifactStore, self).__init__(*args, **kwargs)
+        self._register()
+
+    @staticmethod
+    def open(name: PathType, mode: str = "r") -> Any:
+        """Open a file at the given path."""
+        raise NotImplementedError()
+
+    @staticmethod
+    def copyfile(src: PathType, dst: PathType, overwrite: bool = False) -> None:
```
Comment (ew423) [inline src/coalescenceml/artifact_store/base_artifact_store.py:67]: nitpick, but maybe say what setting overwrite to True/False does? Otherwise really nice docs

File: src/coalescenceml/artifact_store/local_artifact_store.py [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+    def rmtree(path: PathType) -> None:
+        """Deletes dir recursively. Dangerous operation."""
+        shutil.rmtree(path)
+
+    @staticmethod
+    def stat(path: PathType) -> Any:
+        """Return the stat descriptor for a given file path."""
+        return os.stat(path)
+
+    @staticmethod
```
Comment (ew423) [inline src/coalescenceml/artifact_store/local_artifact_store.py:109]: Maybe mention what options do? Dunno if you did doc this elsewhere I'm just going down the files

File: src/coalescenceml/artifacts/type_registry.py [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+if TYPE_CHECKING:
+    from coalescenceml.artifacts.base_artifact import BaseArtifact
+
+logger = get_logger(__name__)
+
+
+class TypeRegistry(object):
+    """Registry to map classes into artifact types"""
+
```
Comment (ew423) [inline src/coalescenceml/artifacts/type_registry.py:20]: mention what options do (unless mentioned in the outward facing side, but it might be worth copy-pasting it in to relevant methods so people know what's going on when working on individual parts of the repo)

File: src/coalescenceml/cli/config.py [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+    profiles = cfg.profiles
+
+    if len(profiles) == 0:
+        cli_utils.warning("No profiles configured!")
+        return
+
+    profile_dicts = []
+    for profile_name, profile in profiles.items():
+        is_active = profile_name == dir_.active_profile_name
+        profile_config = {
```
Comment (ew423) [inline src/coalescenceml/cli/config.py:108]: If you interact with these attributes a lot, it might be worth making an enum for all of these items

File: src/coalescenceml/cli/config.py [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+    """Explains the concept of CoalescenceML profiles."""
+
+    with console.pager():
+        console.print(
+            Markdown(
+                """
+TODO: Explain in words and with examples about what a profile is and the CLI
+command. This should showcase one of every command probably and the most common
+methods that they will want to use.
+"""
```
Comment (ew423) [inline src/coalescenceml/cli/config.py:241]: it do be nice if yall could do this

File: src/coalescenceml/artifact_store/base_artifact_store.py [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+
+    def _register(self, priority: int = 5) -> None:
+        """Create and register a filesystem within the TFX registry"""
+        from tfx.dsl.io.filesystem import Filesystem
+        from tfx.dsl.io.filesystem_registry import DEFAULT_FILESYSTEM_REGISTRY
+
+        filesystem_class = type(
+            self.__class__.__name__,
+            (Filesystem,),
+            {
```
Comment (ew423) [inline src/coalescenceml/artifact_store/base_artifact_store.py:176]: If this list of items is something you reference a lot, may be worth creating an enum

File: src/coalescenceml/cli/config.py [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+            return
+        if cfg.active_profile_name == name:
+            cli_utils.error(
+                f"Profile '{name}' cannot be deleted because it's globally "
+                f" active. Please choose a different active global profile "
+                f"first by running 'coml profile set --global PROFILE'."
+            )
+            return
+
+        if dir_.active_profile_name == name:
```
Comment (ew423) [inline src/coalescenceml/cli/config.py:170]: If there are a lot of strings (e.g. error messages, log notes, etc.), it might be worth storing all of the strings in themed JSON files (e.g. error_messages.json, whatever.json, etc.)

File: src/coalescenceml/cli/stack.py [hunk 16]
```diff
@@ -0,0 +1,341 @@
...
+    """Stacks to define various environments."""
+
+
```
Comment (ew423) [inline src/coalescenceml/cli/stack.py:20]: May be worth making these options their own enum. Not too familiar with enums in Python though - I just know that they're pretty cool in Java - would help clean up this section and make it more readable + provide a centralized location for documentation and ensure consistency across all stack-related things

File: src/coalescenceml/cli/version.py [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+ /:/  \:\  \   /:/  \:\  \   /:/|:|__|__   /:/  /
+/:/__/ \:\__\ /:/__/ \:\__\ /:/ |::::\__\ /:/__/
+\:\  \  \/__/ \:\  \ /:/  / \/__/~~/:/  / \:\  \
+ \:\  \        \:\  /:/  /        /:/  /   \:\  \
+  \:\  \        \:\/:/  /        /:/  /     \:\  \
+   \:\__\        \::/  /        /:/  /       \:\__\
+    \/__/         \/__/         \/__/         \/__/
+    """,
+}
+
```
Comment (ew423) [inline src/coalescenceml/cli/version.py:56]: nice/10

File: src/coalescenceml/config/global_config.py [hunk 18]
```diff
@@ -0,0 +1,453 @@
...
+        environment variable is returned instead.
+        """
+        value = super().__getattribute__(key)
+        if key.startswith("_"):
+            return value
+
+        environment_variable_name = f"{CONFIG_ENV_VAR_PREFIX}{key.upper()}"
+        try:
+            environment_variable_value = os.environ[environment_variable_name]
+            # set the environment variable value to leverage pydantics type
```
Comment (ew423) [inline src/coalescenceml/config/global_config.py:167]: Might be worth making the comments stylistically consistent (e.g. all in triple quotes or all in pound comment or all in ) - I saw some commented code using the pound comment and it could be confusing to see both.

File: src/coalescenceml/constants.py [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+    if value in {"1", "True", "true"}:
+        return True
+    else:
+        return default
+
+
+# Global
+APP_NAME = "CoalescenceML"
+
+# Env variables
```
Comment (ew423) [inline src/coalescenceml/constants.py:41]: Might be worth creating an ENVIRONMENT_VARIABLE enum here to manage all of these var names

File: src/coalescenceml/enums.py [hunk 20]
```diff
@@ -0,0 +1,76 @@
+import logging
+from enum import Enum
+from typing import List
+
```
Comment (ew423) [inline src/coalescenceml/enums.py:4]: yay enums! would love to see more of these :)

File: src/coalescenceml/logger.py [hunk 21]
```diff
@@ -0,0 +1,159 @@
...
+from coalescenceml.constants import APP_NAME, COML_LOGGING_VERBOSITY
+from coalescenceml.enums import LoggingLevels
+
+
+LOG_FILE = f"{APP_NAME}_logs.log"
+
+
+class CustomFormatter(logging.Formatter):
+    """Formats logs according to our specification."""
+
```
Comment (ew423) [inline src/coalescenceml/logger.py:17]: If these are constants, it may be worth making enum/constants/JSON/whatever - generally better to keep hardcoded stuff in one place

File: src/coalescenceml/metadata_store/base_metadata_store.py [hunk 22]
```diff
@@ -0,0 +1,399 @@
...
+                id_=pipeline_context.id,
+                name=pipeline_context.name,
+                metadata_store=self,
+            )
+            pipelines.append(pipeline)
+
+        logger.debug("Fetched %d pipelines.", len(pipelines))
+        return pipelines
+
+    def get_pipeline(self, pipeline_name: str) -> Optional[PipelineView]:
```
Comment (ew423) [inline src/coalescenceml/metadata_store/base_metadata_store.py:198]: may want to rename either this method or the one above - get_pipeline_by_name vs get_pipelines_in_store? so people don't get confused

File: src/coalescenceml/metadata_store/base_metadata_store.py [hunk 23]
```diff
@@ -0,0 +1,399 @@
...
+            return PipelineView(
+                id_=pipeline_context.id,
+                name=pipeline_context.name,
+                metadata_store=self,
+            )
+        else:
+            logger.info("No pipelines found for name '%s'", pipeline_name)
+            return None
+
+    def get_pipeline_runs(
```
Comment (ew423) [inline src/coalescenceml/metadata_store/base_metadata_store.py:214]: maybe rename to get_all_pipeline_runs to differentiate from get_pipeline_run below?

File: src/coalescenceml/orchestrator/utils.py [hunk 24]
```diff
@@ -0,0 +1,282 @@
...
+                        "serialized."
+                    ) from e
+
+    return {key: value for key, value in _inner_generator(obj.dict())}
+
+
+def add_runtime_configuration_to_node(
+    pipeline_node: PipelineNode,
+    runtime_config: RuntimeConfiguration,
+) -> None:
```
Comment (ew423) [inline src/coalescenceml/orchestrator/utils.py:245]: Saw a few methods that modified PipelineNode objects - may be worth separating these methods into their own file

File: src/coalescenceml/pipeline/base_pipeline.py [hunk 25]
```diff
@@ -0,0 +1,432 @@
...
+from coalescenceml.pipeline.schedule import Schedule
+from coalescenceml.stack.exceptions import StackValidationError
+from coalescenceml.step import BaseStep
+from coalescenceml.utils import yaml_utils
+
+
+if TYPE_CHECKING:
+    from coalescenceml.stack import Stack
+
+logger = get_logger(__name__)
```
Comment (ew423) [inline src/coalescenceml/pipeline/base_pipeline.py:46]: May be worth putting these constants in their own constants file/enum/whatever

File: src/coalescenceml/step/utils.py [hunk 26]
```diff
@@ -0,0 +1,502 @@
...
+        return obj
+
+
+class _PropertyDictWrapper(json_utils.Jsonable):
+    """Helper class to wrap inputs/outputs from TFX nodes.
+    Currently, this class is read-only (setting properties is not implemented).
+    Internal class: no backwards compatibility guarantees.
+    Code Credit: https://github.com/tensorflow/tfx/blob
+    /51946061ae3be656f1718a3d62cd47228b89b8f4/tfx/types/node_common.py
+    """
```
Comment (ew423) [inline src/coalescenceml/step/utils.py:103]: May be worth creating a file for this class.

File: src/coalescenceml/step/utils.py [hunk 27]
```diff
@@ -0,0 +1,502 @@
...
+        """Returns the values of the wrapped dictionary."""
+        return self._data.values()
+
+    def items(self) -> ItemsView[str, Channel]:
+        """Returns the items of the wrapped dictionary."""
+        return self._data.items()
+
+
+class _CoMLSimpleComponent(_SimpleComponent):
+    """Simple CoML TFX component with outputs overridden."""
```
Comment (ew423) [inline src/coalescenceml/step/utils.py:160]: Maybe this one too?

**Review by ew423 — COMMENTED** [review]
TL;DR: Really, really nice package. Documentation is incredible for ~0.7 semester's worth of work. Most of my comments are pretty minor nitpicky things that boil down to use enums when there are lots of properties (for things like function options for common functions/object types), using more constant files, and using JSON files to store text outputs for things like error messages/console responses. Basically quarantining hardcoded bits lol. Otherwise, incredible package. Seems really useful! Look forward to seeing what it will become :)

File: src/coalescenceml/step/step_context.py [hunk 28]
```diff
@@ -0,0 +1,156 @@
...
+    from coalescenceml.artifacts.base_artifact import BaseArtifact
+    from coalescenceml.metadata_store.base_metadata_store import (
+        BaseMetadataStore,
+    )
+    from coalescenceml.producers.base_producer import BaseProducer
+    from coalescenceml.stack import Stack
+
+
+class StepContextOutput(NamedTuple):
+    """"""
```
Comment (winnie0617) [inline src/coalescenceml/step/step_context.py:19]: Documentation missing (?)

File: src/coalescenceml/step/step_context.py [hunk 29]
```diff
@@ -0,0 +1,156 @@
...
+
+    producer_class: Type[BaseProducer]
+    artifact: BaseArtifact
+
+
+class StepContext(object):
+    """"""
```
Comment (winnie0617) [inline src/coalescenceml/step/step_context.py:26]: Same here (?)

File: src/coalescenceml/artifact_store/base_artifact_store.py [hunk 30]
```diff
@@ -0,0 +1,196 @@
...
+from coalescenceml.artifact_store.exceptions import ArtifactStoreInterfaceError
+from coalescenceml.enums import StackComponentFlavor
+from coalescenceml.stack import StackComponent
+
+
+PathType = Union[bytes, str]
+
+
+# Bullshit hack b/c of TFX....
+def _catch_not_found_error(_func: Callable[..., Any]) -> Callable[..., Any]:
```
Comment (winnie0617) [inline src/coalescenceml/artifact_store/base_artifact_store.py:28]: I like your detailed documentation for each function and that you specific return type :)

File: src/coalescenceml/io/fileio.py [hunk 31]
```diff
@@ -0,0 +1,373 @@
...
+    "listdir",
+    "makedirs",
+    "mkdir",
+    "open",
+    "remove",
+    "rename",
+    "rmtree",
+    "stat",
+    "walk",
+]
```
Comment (winnie0617) [inline src/coalescenceml/io/fileio.py:373]: Most code is commented out. Is the code still useful?

**Review by winnie0617 — COMMENTED** [review]
It is truly impressive how much you have coded in less than a semester :0 I tried but couldn't really understand fully the architecture so could only give superficial comments :,) But again this looks amazingg
